File: tools/cxx/plan_replay.cpp
// Plans a kept recordings snapshot again through the shipping planner and
// writes what it came to as one JSON object. tools/plan_replay.py builds and
// runs it; see there.

#include "wiiuport/frame/RecordingSnapshot.h"
#include "wiiuport/interp/PlanReplay.h"

#include <array>
#include <cstdint>
#include <cstdio>
#include <exception>
#include <fstream>
#include <iterator>
#include <map>
#include <string>

namespace {

using wiiuport::interp::ObjectPlanner;
using wiiuport::interp::PlanReplay;
using wiiuport::interp::ShaderKey;

std::string readAll(const char* path) {
    std::ifstream in(path, std::ios::binary);
    if (!in) {
        throw std::runtime_error(std::string("cannot read ") + path);
    }
    return {std::istreambuf_iterator<char>(in), std::istreambuf_iterator<char>()};
}

std::string outcomesJson(const ObjectPlanner::Outcomes& outcomes) {
    std::string body = "{";
    for (size_t outcome = 0; outcome < outcomes.size(); ++outcome) {
        body += outcome == 0 ? "\"" : ",\"";
        body += ObjectPlanner::outcomeName(static_cast<ObjectPlanner::Outcome>(outcome));
        body += "\":" + std::to_string(outcomes[outcome]);
    }
    return body + "}";
}

std::string hex(uint64_t value) {
    std::array<char, 17> text{};
    std::snprintf(text.data(), text.size(), "%016llx", static_cast<unsigned long long>(value));
    return text.data();
}

std::string shadersJson(const std::map<ShaderKey, PlanReplay::ShaderReport>& shaders) {
    std::string body = "[";
    for (const auto& [key, shader] : shaders) {
        body += body.size() == 1 ? "{" : ",{";
        body += "\"baseHash\":\"" + hex(key.baseHash) + "\"";
        body += ",\"auxHash\":\"" + hex(key.auxHash) + "\"";
        body += ",\"stageIndex\":" + std::to_string(key.stageIndex);
        body += ",\"outcomes\":" + outcomesJson(shader.outcomes);
        body += ",\"compared\":" + std::to_string(shader.compared);
        body += ",\"leftAtN\":" + std::to_string(shader.leftAtN) + "}";
    }
    return body + "]";
}

std::string reportJson(const PlanReplay::Report& report) {
    std::string body = "{\"frames\":" + std::to_string(report.frames);
    body += ",\"framesPlanned\":" + std::to_string(report.framesPlanned);
    body += ",\"outcomes\":" + outcomesJson(report.outcomes);
    body += ",\"partnersDerived\":" + std::to_string(report.partnersDerived);
    body += ",\"partnersSearched\":" + std::to_string(report.partnersSearched);
    body += ",\"partnersReidentified\":" + std::to_string(report.partnersReidentified);
    body += ",\"reidentifyAttempts\":" + std::to_string(report.reidentifyAttempts);
    body += ",\"partnerCandidates\":" + std::to_string(report.partnerCandidates);
    body += ",\"nearestCandidates\":" + std::to_string(report.nearestCandidates);
    body += ",\"shaders\":" + shadersJson(report.shaders);
    body += ",\"leftAtNByFrame\":[";
    for (size_t frame = 0; frame < report.leftAtNByFrame.size(); ++frame) {
        body += (frame == 0 ? "" : ",") + std::to_string(report.leftAtNByFrame[frame]);
    }
    body += "]";
    body += ",\"planningNanoseconds\":" + std::to_string(report.planning.count()) + "}";
    return body;
}

int replay(int argc, char** argv) {
    if (argc != 3) {
        std::fputs("usage: wiiuport_plan_replay <recordings.bin> <repeats>\n", stderr);
        return 2;
    }
    auto frames = wiiuport::frame::RecordingSnapshot::parse(readAll(argv[1]));
    PlanReplay::Report report = PlanReplay::run(frames, std::stoul(argv[2]));
    std::puts(reportJson(report).c_str());
    return 0;
}

} // namespace

int main(int argc, char** argv) {
    try {
        return replay(argc, argv);
    } catch (const std::exception& failure) {
        std::fprintf(stderr, "wiiuport_plan_replay: %s\n", failure.what());
        return 1;
    }
}

File: tools/wiiuport/planreplay.py
"""What planning a kept snapshot again came to, read from wiiuport_plan_replay.

The executable plans the frames through the shipping planner; this reads its
one JSON line and says what it means per search and per frame, so two versions
of the planner can be compared on the same frames.
"""

from __future__ import annotations

import json
import subprocess
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path

from .cxxtests import build_target
from .paths import Layout

TARGET = "wiiuport_plan_replay"
# Shaders named per ranking: the few that decide what a total means.
RANKED_SHADERS = 5


class PlanReplayFailed(RuntimeError):
    """The snapshot was not planned, so nothing was measured."""


@dataclass(frozen=True)
class ShaderReport:
    """One shader's objects over the planned frames, and what finding them cost."""

    baseHash: str
    auxHash: str
    stageIndex: int
    outcomes: dict[str, int]
    compared: int

    @property
    def name(self) -> str:
        # As the runtime's own reports name a shader.
        return self.baseHash[:8]

    def render(self) -> str:
        objects = sum(self.outcomes.values())
        return (
            f"{self.name}: {objects} objects, {self.outcomes['unverified']} unverified, "
            f"{self.outcomes['unmatched']} unmatched, {self.compared} draws compared"
        )


@dataclass(frozen=True)
class PlanReplayReport:
    frames: int
    framesPlanned: int
    outcomes: dict[str, int]
    partnersDerived: int
    partnersSearched: int
    partnersReidentified: int
    reidentifyAttempts: int
    partnerCandidates: int
    nearestCandidates: int
    shaders: tuple[ShaderReport, ...]
    planningNanoseconds: int

    @classmethod
    def parse(cls, line: str) -> PlanReplayReport:
        fields = json.loads(line)
        fields["shaders"] = tuple(ShaderReport(**shader) for shader in fields["shaders"])
        report = cls(**fields)
        # Planning needs two whole frames before the one planned: a snapshot
        # too short for that measured nothing, however fast it was.
        if report.framesPlanned == 0:
            raise PlanReplayFailed(
                f"none of the {report.frames} frames was planned; a snapshot needs "
                "three frames before any is"
            )
        return report

    def render(self) -> str:
        objects = sum(self.outcomes.values())
        outcomes = ", ".join(f"{name} {count}" for name, count in self.outcomes.items())
        per_frame_ms = self.planningNanoseconds / self.frames / 1e6
        searches = (
            f"{self.partnerCandidates} draws compared for partners over "
            f"{self.partnersSearched} searches "
            f"({self.partnerCandidates / max(1, self.partnersSearched):.1f} each), "
            f"{self.nearestCandidates} for identity over {self.reidentifyAttempts} "
            f"({self.nearestCandidates / max(1, self.reidentifyAttempts):.1f} each)"
        )
        return "\n".join(
            [
                f"planned {self.framesPlanned} of {self.frames} frames: {objects} objects ({outcomes})",
                (
                    f"partners {self.partnersDerived} derived, {self.partnersSearched} searched "
                    f"({self.partnersReidentified} found by their values)"
                ),
                searches,
                f"planning took {per_frame_ms:.3f} ms a frame, the fastest of the repeats",
                f"most compared of {len(self.shaders)} shaders:",
                *self._ranked(lambda shader: shader.compared),
                "most unverified:",
                *self._ranked(lambda shader: shader.outcomes["unverified"]),
            ]
        )

    def _ranked(self, by: Callable[[ShaderReport], int]) -> list[str]:
        ranked = sorted(self.shaders, key=by, reverse=True)[:RANKED_SHADERS]
        return [f"  {shader.render()}" for shader in ranked if by(shader) > 0] or ["  (none)"]


def plan_again(layout: Layout, recordings: Path, repeats: int) -> PlanReplayReport:
    """Build the replay and plan `recordings` through it `repeats` times."""
    if not recordings.is_file():
        raise PlanReplayFailed(f"no recordings snapshot at {recordings}")
    build = build_target(layout, TARGET)
    ran = subprocess.run(
        [str(build / "tools" / "cxx" / TARGET), str(recordings), str(repeats)],
        capture_output=True,
        text=True,
        check=False,
    )
    if ran.returncode != 0:
        raise PlanReplayFailed(f"{TARGET} failed: {(ran.stdout + ran.stderr).strip()}")
    return PlanReplayReport.parse(ran.stdout.strip())
